### PR TITLE
Add option to lookup and run program from path

### PR DESCRIPTION
**Pull request by chrish42**
Uses `runpy.run_path` & `shutil.which`. Implements #97. Any comments on the implementation welcome.

Not sure about the name of the option (`--path` currently), and it would be good to include a test for it too, I assume. Hence marked as draft.

### CODE REVIEW

**Comment by chrish42** [comment]
Added a test, a check for --module and --path being mutually exclusive, and tweaked the argparse help. Let me know what you think. Thanks!

**Comment by chrish42** [comment]
By the way, if you want things to be more uniform with what Python does, it might be relevant to use `runpy.run_path` for the default case too. It is the _other_ function in `runpy`, the complement to `run_module`. It is a Python 3.2+ feature, so it could not be used back when Pyinstrument supported Python 2.7 too. But now that Pyinstrument is supporting only modern Python, that would not be a problem. It would simplify the code too: the only difference with `--path` then would be to actually lookup the program in the $PATH.

I'll let you think if the increase in consistency and decrease in amount of code is worth the potential change in behavior for corner cases...

File: pyinstrument/__main__.py [hunk 1]
```diff
@@ -98,12 +101,15 @@ def dash_m_callback(option, opt, value, parser):
         parser.print_help()
         sys.exit(2)
 
+    if options.module_name is not None and options.path:
+        parser.error("The options --module and --path are mutually exclusive.")
```
[the remark this answers is not in the document]
Comment (chrish42) [reply pyinstrument/__main__.py]: Yes, you are correct. Fixed.

**Comment by chrish42** [comment]
Thanks for the review. Just to be clear, the reason `--from-path` (👍 for the name) is done like that is that I want to transform easily arbitrary invocation of Python programs into profiling run of said programs. With `--from-path` this way, if `program` is a Python program, you can go from running it to profiling it by simply prepending the pyinstrument invocation to the existing command-line:

`program --switch arg1 etc.` -> `pyinstrument -o report.out --other-options --from-path -- program --switch arg1 etc.`

That's the intention. Everything after the `--` is the same. No need to do brittle transformations, shell stuff, etc.

I'd be happy to refactor things to use `run_path` for the "else" branch too and at least have a look at what the tests say, if you're confident that the tests will catch potential problems there / you can have a poke if needed to validate that this wouldn't break things for users overall. If this breaks a lot of tests, I may punt this to another pull request... but my guess is it won't. Just 👍 this comment, and I'll have a look at that.

**Comment by chrish42** [comment]
Change done. Didn't seem to cause problem locally, but I'm not setup to run all the tests. Not sure how to restart the CI run that got interrupted, though. I also added a comment for a change that wasn't clear to me. Please double-check that and generally that the change doesn't break too many things. :-)

**Comment by joerick** [comment]
I made a few changes @chrish42 , mostly more tests to assure myself that we haven't changed anything from the target script's point of view, from our change. I also found some unrelated bugs in the module execution, and fixed them. So this looks good I think. Anything else you'd like to add before I merge?

**Comment by chrish42** [comment]
Looks very good overall. This became a nice cleanup from dropping support for old Python versions, I feel. Thanks for the additional work into my pull request.

The only thing that makes me a little bit sad is `--from-path` not being supported on Windows (even though I wouldn't need it for what I have in mind). What was the issue with there? Not something easy to fix, I'm kinda guessing...?

**Comment by joerick** [comment]
I must admit I'm not an expert in Windows, but it seems like you can't do the POSIX thing of putting a file in PATH, adding a shebang and an executable bit, and it becomes a program. Seems like executables on PATH have to be proper binaries in Windows. This was the error when I tried to run a Python script as an executable on Windows:

```
E                                    OSError: [WinError 216] This version of %1 is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher
```

So by extension, they can't be Python scripts that are profileable by pyinstrument. That said, I do wonder what `pip` does when it installs a package with an entry point - perhaps there's some kind of shim that it installs that invokes Python. But it's not going to work with the same `runpy.run_path` approach that we did for POSIX.

**Comment by chrish42** [comment]
Hmm. Oh, yeah. From memory, `pip` does install some kind of a shim on Windows for scripts. But yeah, script execution on Windows is a bit of a mess. I guess if there's a fix for it, it's probably better handled as a separate pull request.

👍 for merging for me, then.

**Comment by joerick** [comment]
Great! Thanks @chrish42 !

**Pull request merged**